pkg/apis/m3dboperator/v1alpha1/namespace.go

```diff
@@ -98,6 +100,37 @@ type DownsampleOptions struct {
 	All bool `json:"all,omitempty"`
 }
 
+// DynamicOption is an option the type of which is only known at run time.
+type DynamicOption interface {
+	json.Marshaler
+	DeepCopyDynamicOption() DynamicOption
+}
+
+type dynamicOption struct {
+	o interface{}
+}
+
+// NewDynamicOption creates a new DynamicOption.
+func NewDynamicOption(value interface{}) DynamicOption {
+	return &dynamicOption{value}
+}
+
+func (o *dynamicOption) DeepCopyDynamicOption() DynamicOption {
+	// We can fake DeepCopy by resorting to shallow copy here because
+	// within this service dynamic options are effectively immutable.
+	return o
+}
+
+func (o *dynamicOption) MarshalJSON() ([]byte, error) {
+	return json.Marshal(o.o)
+}
+
+// ExtendedOptions stores the extended namespace options.
+type ExtendedOptions struct {
+	Type    string                   `json:"type,omitempty"`
+	Options map[string]DynamicOption `json:"options,omitempty"`
```

Comment:
I've never seen interface types in Kubernetes API types before. I wonder if having this be `map[string]json.RawMessage` would give you the same benefits (auto-generated DeepCopy implementation, more control over serialization) while being more Kubernetes-friendly?

Reply:
Thanks, @schallert , `json.RawMessage` really makes the implementation much simpler. Updated.